cmd/metricchecker/cmd.go

```diff
@@ -0,0 +1,72 @@
+package main
+
+import (
+	"log"
+	"os"
+	"time"
+
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/spf13/cobra"
+
+	"github.com/PingCAP-QE/metrics-checker/pkg/metric"
+)
+
+var rootCmd = &cobra.Command{
+	Use:   "Metrics checker",
+	Short: "For checking prometheus metrics",
+	Run: func(cmd *cobra.Command, args []string) {
+		InitConfig(configFilePath, configBase64)
+
+		address = metric.AddHTTPIfIP(address)
+		if grafanaAPIURL != "" {
+			dashboardName := "Metrics Checker"
+			grafanaAPIURL = metric.AddHTTPIfIP(grafanaAPIURL)
+			err := metric.CreateMetricsDashboard(grafanaAPIURL, dashboardName, config.MetricsToShow)
+			if err != nil {
+				log.Fatalf("Create grafana metrics error: %s", err)
+			}
+			log.Printf("Created dashboard %s on %s", dashboardName, grafanaAPIURL)
+		}
+
+		log.Printf("Start checking metrics after %s", config.StartAfter)
+		for time.Now().Before(config.startTime.Add(config.StartAfter)) {
+			time.Sleep(time.Second)
+		}
+		log.Printf("Start checking metrics")
+		log.Printf("Prometheus address: %s", address)
+
+		// Create prometheus API
+		client, err := api.NewClient(api.Config{
+			Address: address,
+		})
+		if err != nil {
+			log.Fatalf("Create prometheus api failed. address: %s, error: %s", address, err)
+		}
+		api := v1.NewAPI(client)
+
+		for {
+			// run rules
+			for _, rule := range config.Rules {
+				if !metric.Check(api, rule.PromQL, time.Now()) {
+					log.Fatalf("Rule %s failed.", rule)
+				}
+			}
+			time.Sleep(config.Interval)
```

Comment:
I prefer to use a ticker: https://gobyexample.com/tickers

Reply:
Besides, we can put every check in their own goroutines.